proto/marshal/compat.py

```diff
@@ -25,9 +25,17 @@
 except ImportError:
     _message = None
 
+# Try UPB.
+if not _message:
```

Comment:
Please test it more as we discussed. In case if ptotobuf 3.20 and 4.21 re mutually exclusive in this logic, please prioritize workflow, when newer version of protobuf if assumed first, and only if it fails, fallback to older version, not vice versa. This is to make sure that throwing an exception is not an expected behavior for the prioritized protobuf version (which I guess should be the new one)